### ISSUE TITLE
SC2003 should not warn about `expr index` usage

### ISSUE THREAD

**Issue by vapier** [koalaman/shellcheck#1402]
#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
expr index 'abc' b
```

#### Here's what shellcheck currently says:

```
Line 2:
expr index 'abc' b
^-- SC2003: expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].
```

#### Here's what I wanted or expected to see:

It shouldn't warn at all.  There is no builtin shell equivalent for `expr index`.

**Comment by matthewpersico** [comment]
@vapier, I think the point is that `expr` is not antiquated in Bourne shell, that `$((..)), ${}, [[ ]]` are not available in Bourne shell to do arithmetic stuff. And even if `#!/bin/sh` happens to be a link to `bash` at the time when shellcheck is run, you cannot guarantee that going forward; think running shellcheck in dev and then deploying somewhere else.

**Comment by koalaman** [comment]
SC2003 already has a few exceptions for things that are hard to reproduce, like `:`, and `index` could definitely be added to that list.

POSIX says that "The use of string arguments length, substr, index, or match produces unspecified results.", so are there any workarounds that also work while also being completely POSIX?

Maybe `var=abc; prefix=${var%%b*}; echo "${#prefix}"` but you'd have to handle the case of no matching substring.

**Comment by vapier** [comment]
`expr length <str>` can be replaced with `var="<str>"; echo ${#var}` or `echo "<str>" | wc -c`.

`expr substr <string> <pos> <length>` could be replaced with two cuts: `echo "<string>" | cut -b<pos>- | cut -b-<length>`.

`expr match <string> <regex>` (and `expr <string> : <regex>`) could be replaced with `echo "<string>" | grep -q "<regex>"`.

i don't have an answer for `expr index`.  but we should be able to improve SC2003 for the other subcommands in the meantime.

**Comment by chuhn** [comment]
`expr match …` returns matched subexpressions:
```bash
```

**Comment by patsevanton** [comment]
How correct write bash code?
```
echo Execution time was $(expr "$end_time" - "$start_time")  s.
```

**Comment by brother** [comment]
> How correct write bash code?
> 
> ```
> echo Execution time was $(expr "$end_time" - "$start_time")  s.
> ```

either using printf or echo and with bash as destination shell (the thread is about portability though).

```sh
echo "Execution time was $(( end_time - start_time )) s."
printf "Execution time was %d s." $(( end_time - start_time ))
```
